File: src/pedals.py
import digitalio
import time
from buttons import Button
import neopixel
import constants as C

"""Pedal control functionality for DJBB MIDI Loopster 2.0"""

class FakeKeypadEvent:
    """A fake keypad event that mimics CircuitPython's keypad event structure"""
    def __init__(self, key_number, pressed):
        self.key_number = key_number
        self.pressed = pressed
        self.released = not pressed

class Pedals:
    def __init__(self):
        self.pixels = neopixel.NeoPixel(C.PEDAL_NEOPIXEL_PIN, C.PEDAL_COUNT, brightness=C.PEDAL_BRIGHTNESS, auto_write=False)
        self.pixels.fill((0, 0, 0))
        self._pedal_1 = digitalio.DigitalInOut(C.PEDAL_1_PIN)
        self._pedal_1.direction = digitalio.Direction.INPUT
        self._pedal_1.pull = digitalio.Pull.UP

        self._pedal_2 = digitalio.DigitalInOut(C.PEDAL_2_PIN)
        self._pedal_2.direction = digitalio.Direction.INPUT
        self._pedal_2.pull = digitalio.Pull.UP

        self._pedal_3 = digitalio.DigitalInOut(C.PEDAL_3_PIN)
        self._pedal_3.direction = digitalio.Direction.INPUT
        self._pedal_3.pull = digitalio.Pull.UP
        
        self._pedal_4 = digitalio.DigitalInOut(C.PEDAL_4_PIN)
        self._pedal_4.direction = digitalio.Direction.INPUT
        self._pedal_4.pull = digitalio.Pull.UP
        
        self._pedal_5 = digitalio.DigitalInOut(C.PEDAL_5_PIN)
        self._pedal_5.direction = digitalio.Direction.INPUT
        self._pedal_5.pull = digitalio.Pull.UP

        self.pedal_1 = Button(pad_index=0,label="Pedal 1")
        self.pedal_2 = Button(pad_index=1,label="Pedal 2")
        self.pedal_3 = Button(pad_index=2,label="Pedal 3")
        self.pedal_4 = Button(pad_index=3,label="Pedal 4")
        self.pedal_5 = Button(pad_index=4,label="Pedal 5")
        self.pedals = [self.pedal_1, self.pedal_2, self.pedal_3, self.pedal_4, self.pedal_5]

        self.pixels_need_update = True

        # Event queue for fake keypad events
        self.event_queue = []
        
        # Bank state variables
        self.current_bank = 0  # 0, 1, or 2 for banks 0-2
        self.bank_offset = 0   # Calculated offset: bank * 5

    def update(self):
        """Update the state of all pedals and generate fake keypad events"""
        self.pedal_1.set_current_value(self._pedal_1.value)
        self.pedal_2.set_current_value(self._pedal_2.value)
        self.pedal_3.set_current_value(self._pedal_3.value)
        self.pedal_4.set_current_value(self._pedal_4.value)
        self.pedal_5.set_current_value(self._pedal_5.value)
        
        for i, pedal in enumerate(self.pedals):
            # Store previous state before updating
            prev_state = pedal.state
            pedal.update_all()
            
            # Generate fake events for state changes
            if not prev_state and pedal.state:  # New press
                fake_event = FakeKeypadEvent(key_number=pedal.pad_idx, pressed=True)
                self.event_queue.append(fake_event)
            elif prev_state and not pedal.state:  # New release
                fake_event = FakeKeypadEvent(key_number=pedal.pad_idx, pressed=False)
                self.event_queue.append(fake_event)
    
    def show_pedal_pixels(self):
        """Update the pedal pixels based on the current pedal states"""
        if self.pixels_need_update:
            self.pixels.show()
            self.pixels_need_update = False

    def set_pixel_on(self, loopster_pad_idx, color=C.NOTE_COLOR):
        """Set the pixel color for a pedal based on loopster pad index"""
        pedal_pixel_idx = self.get_pedal_pixel_index(loopster_pad_idx)
        if pedal_pixel_idx is not None:
            self.pixels[pedal_pixel_idx] = color
            self.pixels_need_update = True

    def set_pixel_off(self, loopster_pad_idx):
        """Turn off the pixel for a pedal based on loopster pad index"""
        pedal_pixel_idx = self.get_pedal_pixel_index(loopster_pad_idx)
        if pedal_pixel_idx is not None:
            self.pixels[pedal_pixel_idx] = (0, 0, 0)
            self.pixels_need_update = True
    
    def reset_pedals(self):
        """Reset the pedal states and pixel colors"""
        for pedal in self.pedals:
            pedal.reset_actions()

    def get_event(self):
        """Get the next event from the pedal event queue, returns None if empty"""
        if self.event_queue:
            return self.event_queue.pop(0)
        return None

    def set_pedal_bank(self, bank_idx):
        """Switch to a different pedal bank (0-2)"""
        if 0 <= bank_idx <= 2:
            self.current_bank = bank_idx
            self.bank_offset = bank_idx * 5
            # Update all pedal button pad_indices
            for i, pedal in enumerate(self.pedals):
                pedal.pad_idx = self.bank_offset + i
            
            # Immediately reflect the state of loops in the new bank
            self._update_pixels_for_current_bank()
            
            print(f"Switched to pedal bank {bank_idx} (pads {self.bank_offset}-{self.bank_offset+4})")

    def get_pedal_pixel_index(self, loopster_pad_idx):
        """Convert loopster pad index to local pedal pixel index (0-4)"""
        if self.bank_offset <= loopster_pad_idx < self.bank_offset + 5:
            return loopster_pad_idx - self.bank_offset
        return None  # This pad doesn't correspond to any pedal

    def _update_pixels_for_current_bank(self):
        """Update pedal pixels to reflect current state of loops in this bank"""
        # Late import to avoid circular dependency
        from loopmanager import loop_manager
        
        # For each pedal in current bank, check if corresponding loop is playing
        for i in range(5):
            loopster_pad_idx = self.bank_offset + i
            # Check if there's a loop at this pad index and if it's playing
            if (loopster_pad_idx < len(loop_manager.loops) and 
                loop_manager.loops[loopster_pad_idx] != "" and 
                loop_manager.loops[loopster_pad_idx].loop_is_playing):
                self.pixels[i] = C.PIXEL_LOOP_PLAYING_COLOR
            elif (loopster_pad_idx < len(loop_manager.loops) and 
                  loop_manager.loops[loopster_pad_idx] != ""):
                # Has a loop but not playing - use LOOP_COLOR
                self.pixels[i] = C.LOOP_COLOR
            else:
                # No loop
                self.pixels[i] = (0, 0, 0)
        self.pixels_need_update = True

# Create the global pedals instance
pedals = Pedals()

def update_pedal_pixels():
    """Update the pedal pixels based on the current pedal states"""
    if pedals.pixels_need_update:
        pedals.show_pedal_pixels()

File: src/pedals.cpp
#include "pedals.h"
#include "pixels.h" // blink_phase + core1_owns_pixels
#include "serial_config.h" // cdc_log (Q1 bounded prints)

Pedals pedals;

const uint8_t Pedals::_pedal_pins[C::PEDAL_COUNT] = {
    C::PIN_PEDAL_1, C::PIN_PEDAL_2, C::PIN_PEDAL_3, C::PIN_PEDAL_4, C::PIN_PEDAL_5
};

Pedals::Pedals()
    : _pixels(C::PEDAL_COUNT, C::PIN_PEDAL_NEOPIXEL, NEO_GRB + NEO_KHZ800) {}

void Pedals::begin() {
    _pixels.begin();
    _pixels.setBrightness((uint8_t)(C::PEDAL_BRIGHTNESS * 255));
    _pixels.clear();
    for (uint8_t i = 0; i < C::PEDAL_COUNT; i++) {
        pinMode(_pedal_pins[i], INPUT_PULLUP);
    }
}

void Pedals::_push_event(uint8_t key, bool pressed) {
    constexpr size_t N = sizeof(_queue) / sizeof(_queue[0]);
    if (_q_count >= N) {
        // Full: drop the OLDEST, never the incoming event. Dropping the newest could eat a
        // RELEASE, stranding note_buttons[].state as held — stuck note plus the encoder
        // routed to the pad-held handler until reboot. (Bounded vs Python's unbounded list.)
        _q_head = (_q_head + 1) % N;
        _q_count--;
    }
    _queue[(_q_head + _q_count) % N] = {key, pressed};
    _q_count++;
}

bool Pedals::get_event(PedalEvent &out) {
    if (_q_count == 0) {
        return false;
    }
    out = _queue[_q_head];
    _q_head = (_q_head + 1) % (sizeof(_queue) / sizeof(_queue[0]));
    _q_count--;
    return true;
}

void Pedals::update() {
    for (uint8_t i = 0; i < C::PEDAL_COUNT; i++) {
        pedal_buttons[i].set_current_value(digitalRead(_pedal_pins[i]) == HIGH);

        bool prev_state = pedal_buttons[i].state;
        pedal_buttons[i].update_all();

        if (!prev_state && pedal_buttons[i].state) { // new press
            _push_event((uint8_t)pedal_buttons[i].pad_idx, true);
        } else if (prev_state && !pedal_buttons[i].state) { // new release
            _push_event((uint8_t)pedal_buttons[i].pad_idx, false);
        }
    }
}

void Pedals::show_pedal_pixels() {
    if (pixels_need_update) {
        pixels_need_update = false; // clear first: a concurrent core-0 write re-dirties
        _pixels.show();
    }
}

void Pedals::reset_pedals() {
    for (uint8_t i = 0; i < C::PEDAL_COUNT; i++) {
        pedal_buttons[i].reset_actions();
    }
}

void Pedals::set_pedal_bank(int bank_idx) {
    if (bank_idx < 0 || bank_idx > 2) {
        return;
    }
    current_bank = bank_idx;
    bank_offset = bank_idx * 5;
    for (uint8_t i = 0; i < C::PEDAL_COUNT; i++) {
        // A pedal held across the bank change would otherwise deliver its release under the
        // NEW pad index: the old pad never sees the release, so it stays "held" forever —
        // stuck note, and every later encoder turn is routed to the pad-held handler
        // instead of menu nav. Close the press out on the index it opened on first.
        if (pedal_buttons[i].state) {
            _push_event((uint8_t)pedal_buttons[i].pad_idx, false);
        }
        pedal_buttons[i].pad_idx = bank_offset + i;
    }
    _update_pixels_for_current_bank();
    cdc_log("Switched to pedal bank %d (pads %d-%d)\n", bank_idx, bank_offset, bank_offset + 4);
}

int Pedals::get_pedal_pixel_index(uint8_t loopster_pad_idx) const {
    if (loopster_pad_idx >= bank_offset && loopster_pad_idx < bank_offset + 5) {
        return loopster_pad_idx - bank_offset;
    }
    return -1; // this pad doesn't correspond to any pedal
}

C::Rgb Pedals::_color_for(PadLoopState st, bool blink_phase) {
    switch (st) {
    case PadLoopState::Playing:
        return C::PIXEL_LOOP_PLAYING_COLOR;
    case PadLoopState::HasLoop:
        return C::LOOP_COLOR;
    case PadLoopState::Queued:
        return blink_phase ? C::PIXEL_LOOP_PLAYING_COLOR : C::BLACK;
    case PadLoopState::Recording:
        return C::RED;
    case PadLoopState::Armed:
        return blink_phase ? C::RED : C::BLACK;
    default:
        return C::BLACK;
    }
}

void Pedals::refresh_from_loop_state() {
    if (!_loop_state) {
        return; // loopmanager not initialized yet
    }
    // pixels.blink_phase only advances while some PAD pixel blinks; every Queued/Armed
    // pad blinks on the pad strip too, so the phase is always live when we need it.
    bool phase = pixels.blink_phase;
    bool phase_changed = (phase != _last_phase) || !_rendered_once;
    bool dirty = false;
    for (uint8_t i = 0; i < C::PEDAL_COUNT; i++) {
        PadLoopState st = _loop_state((uint8_t)(bank_offset + i));
        bool blinks = (st == PadLoopState::Queued || st == PadLoopState::Armed);
        if (st == _last_state[i] && _rendered_once && !(blinks && phase_changed)) {
            continue;
        }
        C::Rgb c = _color_for(st, phase);
        _pixels.setPixelColor(i, c.r, c.g, c.b);
        _last_state[i] = st;
        dirty = true;
    }
    _last_phase = phase;
    _rendered_once = true;
    if (dirty) {
        pixels_need_update = true;
    }
}

void Pedals::_update_pixels_for_current_bank() {
    _rendered_once = false; // bank changed: every LED must be re-derived
    refresh_from_loop_state();
}

void update_pedal_pixels() {
    if (pixels.core1_owns_pixels) {
        return; // core 1 shows pedal pixels from loop1; a core-0 show would race it
    }
    if (pedals.pixels_need_update) {
        pedals.show_pedal_pixels();
    }
}
